# 1. temp. conversion

In [5]:
def tempe(temp,unit):
    if unit == "C".lower():
        return temp * 9/5 + 32 #c to f

    elif unit == "F".lower():
        return (temp-32)*5/9 # f to c
    else:
        return f"{temp} {unit} not supported"

tempe(25,"a")

'25 a not supported'

# Problem Explanation

Converts temperature between Celsius and Fahrenheit units.

# Approach Used

* **Technique:** Conditionals and formulas.
* **Why it works:** Evaluates the target unit and applies conversion formulas: `(C * 9/5) + 32` or `(F - 32) * 5/9`.
* **Key Idea:** Unit mapping using checks.

# Dry Run

Input: `tempe(25, "a")`
* Unit is "a". Converts to lower: "a".
* Matches neither "c" nor "f".
* Returns `"25 a not supported"`.

# Time Complexity

* **Time:** O(1)
* **Explanation:** Conditional statements run in constant time.

# Space Complexity

* **Space:** O(1)
* **Explanation:** Simple variable values.


# 2. password strrengh checker


In [15]:
def is_pass_strong(password):
    if len(password)<8:
        return False
    if not any(char.isdigit() for char in password):
        return False
    if not any(char.islower() for char in password):
        return False
    if not any(char.isupper() for char in password):
        return False
    if not any(char in '!@#$%^&*()_+' for char in password):
        return False
    return True
print(is_pass_strong("Vaidik@8799"))

True


# Problem Explanation

Validates password strength by verifying minimum length and the presence of digits, lowercase, uppercase, and special characters.

# Approach Used

* **Technique:** String Validation via generators.
* **Why it works:** Uses `any()` to ensure that at least one character in the string satisfies each respective criteria.
* **Key Idea:** Multi-constraint check on character classes.

# Dry Run

Input: `"Vaidik@8799"`
* Length is 11 (>= 8) -> Pass.
* Contains digits? Yes ('8', '7', etc.) -> Pass.
* Contains lowercase? Yes ('a', 'i', etc.) -> Pass.
* Contains uppercase? Yes ('V') -> Pass.
* Contains special? Yes ('@') -> Pass.
* Returns `True`.

# Time Complexity

* **Time:** O(L)
* **Explanation:** Where `L` is the password length; iterates character string up to 5 times.

# Space Complexity

* **Space:** O(1)
* **Explanation:** No extra data structures.


# 3 . calclulating total cost of item from cart

In [19]:
## sample data
cart = [
    {'name':'apple','price':0.5,'qun':6},
    {'name':'mango','price':0.7,'qun':5},
    {'name':'graps','price':0.2,'qun':8},

]

def total_item_cost(cart):
    
    total_cost = 0
    for item in cart:
        total_cost+= item['price'] * item['qun']
    print(total_cost)

#calling function

total_item_cost(cart)

8.1


# Problem Explanation

Calculates the total cost of items in a shopping cart list of dictionaries.

# Approach Used

* **Technique:** Loop accumulation.
* **Why it works:** Iterates over each item, multiplies its price by quantity, and adds to a running sum.
* **Key Idea:** Running sum of products.

# Dry Run

Input: `cart` list.
* `total_cost = 0`
* Item 1: `apple`. Price 0.5 * Qty 6 = 3.0. `total_cost` = 3.0.
* Item 2: `mango`. Price 0.7 * Qty 5 = 3.5. `total_cost` = 6.5.
* Item 3: `graps`. Price 0.2 * Qty 8 = 1.6. `total_cost` = 8.1.
* Prints `8.1`.

# Time Complexity

* **Time:** O(N)
* **Explanation:** Where `N` is the number of items in the cart.

# Space Complexity

* **Space:** O(1)
* **Explanation:** Running total accumulator.


# 4. palinedrome string
<p><h5>same from start and end both <h5><p>

In [27]:
def pali(s):
    s=s.lower().replace(" ","")
    if s == s[::-1]:
        print(f"{s} is palindrome")
    else:
        print(f"{s} is not palindrome")
pali("No lemon no melon")

nolemonnomelon is palindrome


# Problem Explanation

Checks if a string is a palindrome by comparing it to its reverse, ignoring whitespace and letter casing.

# Approach Used

* **Technique:** String Slicing and reversal comparison.
* **Why it works:** Cleans string by replacing spaces with empty string and converting to lowercase, then checks if it is equal to `s[::-1]`.
* **Key Idea:** Clean string comparison with reverse.

# Dry Run

Input: `"No lemon no melon"`
* Lowercased & replaced: `"nolemonnomelon"`.
* Reverse is `"nolemonnomelon"`.
* String equals reverse, prints: `nolemonnomelon is palindrome`.

# Time Complexity

* **Time:** O(N)
* **Explanation:** Reversing and cleaning string take linear time.

# Space Complexity

* **Space:** O(N)
* **Explanation:** To store clean string and reversed copy.


# 5. factorial of number using recursion

In [32]:
#exa,ple = 5! = 5*4*3*2*1
def factorial(n):
    if n == 0:
        return 1
    else:
        return n * factorial(n-1)

print(factorial(6))


720


# Problem Explanation

Computes the factorial of a number using recursion.

# Approach Used

* **Technique:** Recursion.
* **Why it works:** Leverages the mathematical definition `n! = n * (n-1)!` with a base case at `n = 0`.
* **Key Idea:** Breakdown of subproblems via recursive call stacks.

# Dry Run

Input: `factorial(6)`
* `factorial(6) = 6 * factorial(5)`
* `factorial(5) = 5 * factorial(4)`
* ...
* `factorial(0)` returns `1`.
* Reconstructs: `1 * 1 * 2 * 3 * 4 * 5 * 6 = 720`.

# Time Complexity

* **Time:** O(N)
* **Explanation:** `N` recursive stack calls.

# Space Complexity

* **Space:** O(N)
* **Explanation:** Call stack depth of `N`.


# 6. read file and count freq. of each words

In [45]:
def count_freq_word(file_path):
    word_count={}
    with open(file_path, 'r') as file:
        for line in file:
            words = line.split()
            for word in words:
                word = word.lower().strip('!.,:;"\'')
                word_count[word]=word_count.get(word,0)+1
    return word_count
file_path = r"sample.txt"
print(count_freq_word(file_path) )




{'namaste': 1, 'bhai': 1, 'log': 1, 'kaise': 1, 'ho': 1, '': 1, 'kya': 1, 'hall': 1, 'he?': 1}


# Problem Explanation

Reads a text file and counts the frequency of each unique word, ignoring case and stripping punctuation marks.

# Approach Used

* **Technique:** Dictionary Mapping with File I/O.
* **Why it works:** Reads line by line, splits into tokens, cleans each token, and increments its count in a hash map.
* **Key Idea:** Text cleaning and word count tracking.

# Dry Run

Input: File with text `"Hello World. Hello."`
* Word `"Hello."` cleaned to `"hello"`, map: `{'hello': 1}`.
* Word `"World."` cleaned to `"world"`, map: `{'hello': 1, 'world': 1}`.
* Word `"Hello."` cleaned to `"hello"`, map: `{'hello': 2, 'world': 1}`.
* Returns word counts dict.

# Time Complexity

* **Time:** O(W)
* **Explanation:** Where `W` is the total number of words in the file.

# Space Complexity

* **Space:** O(U)
* **Explanation:** Where `U` is the number of unique words.


# 7 . validate email

In [47]:
import re
def email_chk(email):
    pattern = r'^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$'
    if re.match(pattern, email):
        print(f"{email} is valid")
    else:
        print(f"{email} is not valid")

email_chk("vaidikpipaliya@gmail.com")

vaidikpipaliya@gmail.com is valid


# Problem Explanation

Validates if a given email string matches standard email formatting constraints using regular expressions.

# Approach Used

* **Technique:** Regex validation.
* **Why it works:** Match against a pattern checking local name, `@` separator, domain name, and dot suffix extension.
* **Key Idea:** Regex string matching.

# Dry Run

Input: `"vaidikpipaliya@gmail.com"`
* Pattern: `^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$`
* Matches successfully.
* Prints: `vaidikpipaliya@gmail.com is valid`.

# Time Complexity

* **Time:** O(L)
* **Explanation:** Regex matching is linear with respect to email length `L`.

# Space Complexity

* **Space:** O(1)
* **Explanation:** Minimal regex state engine memory.
